scripts/sign.py

```diff
@@ -63,16 +59,45 @@ def main():
     h.update(shdr_uuid)
     h.update(shdr_version)
     h.update(img)
-    sig = signer.sign(h)
-
-    f = open(args.out, 'wb')
-    f.write(shdr)
-    f.write(h.digest())
-    f.write(sig)
-    f.write(shdr_uuid)
-    f.write(shdr_version)
-    f.write(img)
-    f.close()
+    img_digest = h.digest()
+
+    def write_image_with_signature(sig):
+        with open(args.out, 'wb') as f:
+            f.write(shdr)
+            f.write(img_digest)
+            f.write(sig)
+            f.write(shdr_uuid)
+            f.write(shdr_version)
+            f.write(img)
+
+    if key.has_private():
+        signer = PKCS1_v1_5.new(key)
+        sig = signer.sign(h)
+        write_image_with_signature(sig)
+    else:
+        print('Provided key cannot be used for signing, switching to offline-signing mode.')
```

Comment:
Yes, I'm with you on reducing the verbosity (and probably looking at the V-level from the environment.

Regarding documentation, I would put it into documentation/optee_design.md (but please take care https://github.com/OP-TEE/optee_os/pull/2043 first).

Regarding API:
From a security POV the script should probably bail out and run around screaming, when it encounters a private key (in the source tree). But then again, for development purposes, we might accept this. Offline signing with the private key present violates the definition of offlineand and supporting this practice is like preparing the gun for the user shooting his own foot (later).
Aside from that, I'm rather against introducing new verbs. There are three reasonable setting that the script may encounter and for each one, there is exactly one sensible behaviour. New verbs would imposer learning the verbs on the user, checking for them in the script and generating error messages when the situation found by the script does not match the situation expected by the verb.
In the proposed form, the script just works. When a developer switches to offline-signing, she will surely appreciate when change is as simple as:
1. Replace Key with public-only-key
2. make
3. sign offline
4. make

Once offline signing is tested and going smooth, she can adopt her Makefile easily without looking up further documentation.

That said, I hope there's more than the two of us reading these comments, so we can have some sort of popular vote. (I'm not religious about my solution, I just personnaly prefer it.)

Reply:
> I'm with you on reducing the verbosity (and probably looking at the V-level from the environment.

Sounds good.

> Regarding documentation, I would put it into documentation/optee_design.md 

Agreed in principle but I suppose it would be time to switch to the new Git: https://github.com/OP-TEE/optee_docs

> (but please take care #2043 first).

No, *you* please take care of @jenswi-linaro's comment: "Please rebase to latest and squash in the fixups in their commits so we can continue this review." :wink:

> That said, I hope there's more than the two of us reading these comments, so we can have some sort of popular vote. (I'm not religious about my solution, I just personnaly prefer it.)

+1

I am all for using simple solutions as long as they do not introduce limitations we may regret later. As you said, if your proposal fits the expected use cases well without making undesirable assumptions, then it is probably fine. And this is where I am not convinced. You are assuming hard-coded names instead of taking names as arguments, and the presence or absence of a file changes the behavior. IMO, those are things that should be taken care of outside of the script itself (Makefile typically). Compare with the `openssl` command for instance.

Reply:
> Agreed in principle but I suppose it would be time to switch to the new Git: https://github.com/OP-TEE/optee_docs

It will be a honor to make one of the first PRs there. :)
 
> I am all for using simple solutions as long as they do not introduce limitations we may regret later. As you said, if your proposal fits the expected use cases well without making undesirable assumptions, then it is probably fine. And this is where I am not convinced. You are assuming hard-coded names instead of taking names as arguments, and the presence or absence of a file changes the behavior. IMO, those are things that should be taken care of outside of the script itself (Makefile typically). Compare with the `openssl` command for instance.

How about this: 
There will be three commands: auto | generate_digest | stitch_ta with auto being the default :)
I'll also add options to override hardcoded filenames.

Reply:
Back to your example:

> In the proposed form, the script just works. When a developer switches to offline-signing, she will surely appreciate when change is as simple as:
> 1. Replace Key with public-only-key
> 2. make
> 3. sign offline
> 4. make

I understand it looks convenient but I think you have not realized the drawbacks and limitations. IMO in order to support direct and offline signing, the Makefile has to be changed anyways. This cannot be hidden in `sign.py` alone. The reason being, `make` should not have a different meaning depending on whether the private key is present locally or not (steps 2 and 4). In Makefiles, the dependencies have to be clearly defined, and rules  should have some well known input and output.

In the current state of the code, step 2 above is supposed to invoke the script to generate the `.ta` files. With your proposal and without the private key, it can not, and will generate a digest instead. So if you type `make` again, it will re-run the script. And still no `.ta` file will be produced. First bad thing.

Then suppose you want to extend the Makefile to do something with the `.ta` files, such as adding some packaging rules. How would you do that?

Adding a configuration flag and an additional target is a clean solution to this problem. For instance:
```make
all: ta

ta: a.ta

ifdef OFFLINE
a.ta: a.elf a.sig
	sign.py stitch_ta ...

digests: a.dig

a.dig: a.elf
	sign.py generate_digest ...

else
a.ta: a.stripped.elf privkey.pem
	sign.py sign ...    # "sign" action could be omitted (be the default)
endif

a.elf:
	$(CC) ...

package: ta
	tar cf ...
```

Usage would be:
```
# Direct signing with privkey.pem
$ make
$ make package

# Offline signing, privkey.pem not used at all
$ export OFFLINE=1
$ make digests
$ openssl ... # Generate a.sig
$ make
$ make package
```

At this point it is easy to invoke the script with the proper verb and arguments. Granted, this does not strictly mandate the use of additional verbs, because the action can be deduced from the arguments (is `--key` present? is `--uuid` present?). But it sounds simpler to have a verb and error out if some argument is missing.

Finally, if for some reason you find it useful to support the semi-broken use case you mentioned in steps 1-4, you can easily use a wrapper script.

Reply:
You're right. I didn't think far enough regarding the semantics of make. When I call a (sub)target `foo.ta`, I expect to get an artefact `foo.ta` (no PHONY targets here). If not, something must have gone wrong. Getting a digest file instead does not indicate success and calling the target again, make will simply detect a missing `foo.ta` and start rebuilding stuff. Which is not what I want it to do, but what I tell it to do.
I'll rework the script asap and update the PR.

Reply:
> I'll rework the script asap and update the PR.

Great, thanks. Do not care too much about the ***asap*** part of it and do not hesitate to ping after few days without feedback.

Reply:
I think this is resolved already.